app/controllers/ActionBuilders.scala

```diff
@@ -55,20 +58,51 @@ trait ActionBuilders {
     }
   }
 
-  def fieldDefinitionsExistRefiner(implicit ec: ExecutionContext): ActionRefiner[ApplicationRequest, ApplicationWithFieldDefinitionsRequest]
+  def fieldDefinitionsExistRefiner(noFieldsBehaviour : NoSubscriptionFieldsRefinerBehaviour)
+                                  (implicit ec: ExecutionContext): ActionRefiner[ApplicationRequest, ApplicationWithFieldDefinitionsRequest]
       = new ActionRefiner[ApplicationRequest, ApplicationWithFieldDefinitionsRequest] {
 
     def refine[A](input: ApplicationRequest[A]): Future[Either[Result, ApplicationWithFieldDefinitionsRequest[A]]] = {
       implicit val implicitRequest: Request[A] = input.request
 
+      val noFieldsResult = noFieldsBehaviour match {
+        case NoSubscriptionFieldsRefinerBehaviour.BadRequest => play.api.mvc.Results.NotFound(errorHandler.notFoundTemplate)
+        case NoSubscriptionFieldsRefinerBehaviour.Redirect(url) => play.api.mvc.Results.Redirect(url)
+      }
+
+      val apiSubscriptionStatuses =
+        input.subscriptions.filter(s => s.subscribed)
+
+      val apiSubStatusesWithFieldDefinitions = NonEmptyList
+        .fromList(APISubscriptionStatusWithSubscriptionFields(apiSubscriptionStatuses).toList)
+
+      Future.successful(
+        apiSubStatusesWithFieldDefinitions
+          .fold[Either[Result, ApplicationWithFieldDefinitionsRequest[A]]]
+            (Left(noFieldsResult))
+            (withDefinitions => Right(ApplicationWithFieldDefinitionsRequest(withDefinitions, input)))
+      )
+    }
+  }
+
+  def subscriptionFieldPageRefiner(pageNumber: Int)(implicit ec: ExecutionContext):
+    ActionRefiner[ApplicationWithFieldDefinitionsRequest, ApplicationWithSubscriptionFieldPage]
+      = new ActionRefiner[ApplicationWithFieldDefinitionsRequest, ApplicationWithSubscriptionFieldPage] {
+
+    def refine[A](input: ApplicationWithFieldDefinitionsRequest[A]): Future[Either[Result, ApplicationWithSubscriptionFieldPage[A]]] = {
+      implicit val implicitRequest: Request[A] = input.applicationRequest.request
+
+      val details = input.fieldDefinitions.map(toDetails).toList
+
       Future.successful(
-        NonEmptyList.fromList(
-          input.subscriptions.filter(
-            s => s.subscribed && s.fields.isDefined
-          ).toList
-        )
-        .map(nel => ApplicationWithFieldDefinitionsRequest(nel, input))
-        .toRight(play.api.mvc.Results.NotFound(errorHandler.notFoundTemplate))
+        if (pageNumber >= 1 && pageNumber <= details.size) {
```

Comment:
So who wants to make pageNumber a refined type of positive integer ?

Reply:
Yeah given that is passed in from the URL it needs to be safe.

Reply:
I assume you mean with this? https://github.com/fthomas/refined
It's not available for Scala 2.11 according to their docs. I've just tried it anyway and it fails to find a 2.11 version.